R/shiny.R

```diff
@@ -0,0 +1,307 @@
+#' @export
+require_oauth <- function(app, oauth_app, scopes, welcome_ui,
+  cookie_opts = cookie_options(http_only = TRUE)) {
+
+  force(oauth_app)
+  force(scopes)
+  force(welcome_ui)
+
+  httpHandler <- app$httpHandler
+  app$httpHandler <- function(req) {
+    resp <-
+      handle_oauth_callback(req, oauth_app, cookie_opts) %||%
+      handle_logged_in(req, oauth_app, httpHandler) %||%
+      handle_welcome(req, oauth_app, scopes, cookie_opts)
+    resp
+  }
+
+  serverFuncSource <- app$serverFuncSource
+  app$serverFuncSource <- function() {
+    wrappedServer <- serverFuncSource()
+    function(input, output, session) {
+      creds <- read_creds_from_cookies(session$request, oauth_app)
+      if (is.null(creds)) {
+        stop("gargle_token cookie expected but not found")
+      } else {
+        session$userData$gargle_token <- creds
+        wrappedServer(input, output, session)
+      }
+    }
+  }
+
+  app
+}
+
+handle_oauth_callback <- function(req, oauth_app, cookie_opts) {
+  if (has_code_param(req)) {
+    # User just completed login; verify, set cookie, and redirect
+    cookies <- parse_cookies(req)
+    gargle_auth_state <- cookies[["gargle_auth_state"]]
+    if (!is.null(gargle_auth_state)) {
+      qs <- shiny::parseQueryString(req[["QUERY_STRING"]])
+      code <- qs$code
+      state <- qs$state
+
+      if (identical(state, gargle_auth_state)) {
+        cred <- httr::oauth2.0_access_token(
+          gargle_outh_endpoint(),
+          app = oauth_app,
+          code = code,
+          redirect_uri = infer_app_url(req)
+        )
+
+        # cred has:
+        # access_token, expires_in, scope, token_type, and id_token
+        # (and possibly refresh_token)
+
+        return(shiny::httpResponse(
+          status = 307L,
+          content_type = "text/plain",
+          content = "",
+          headers = rlang::list2(
+            Location = infer_app_url(req),
+            "Cache-Control" = "no-store",
+            !!!delete_cookie_header("gargle_auth_state", cookie_opts),
+            !!!set_cookie_header("gargle_token", wrap_creds(cred, oauth_app),
+              cookie_opts)
+          )
+        ))
+      }
+    }
+  }
+}
+
+handle_logged_in <- function(req, oauth_app, httpHandler) {
+  if (!is.null(read_creds_from_cookies(req, oauth_app))) {
+    # User is already logged in, proceed
+    return(httpHandler(req))
+  }
+}
+
+handle_welcome <- function(req, oauth_app, scopes, cookie_opts) {
+  redirect_uri <- infer_app_url(req)
+  state <- sodium::bin2hex(sodium::random(32))
+  query_extra <- list(
+    access_type = "offline"
+  )
+
+  # TODO: Add email?
+
+  auth_url <- httr::oauth2.0_authorize_url(
+    endpoint = gargle_outh_endpoint(),
```

Comment:
```suggestion
    endpoint = gargle_oauth_endpoint(),
```

Reply:
Actually the way I have it is "correct" 😬 

I'm happy to fix the actual function in a separate PR if you like?

Reply:
😱

Reply:
Fixed in 848663bf7e930bf59f69956908997d782fecee2f, so you can merge/rebase